frontend/src/components/placementVisuals/Placement.tsx

```diff
@@ -242,6 +248,8 @@ const Placement = ({ sourceSamples, destinationSamples, cycleContainer, saveChan
         updateSampleList(samplesToUpdate, type)
     }, [selectedSamples, sourceSamples.samples, destinationSamples.samples])
 
+    const disableUndo = useMemo(() => Object.values(selectedSamples).some(sample => sample.type == 'source'), [selectedSamples])
```

Comment:
Whenever you have those harcoded values, you should try to import the corresponding constant (SOURCE_STRING here).